README.md
# Strictly Speaking

[![Build Status](https://github.com/gmile/strictly_speaking/actions/workflows/ci.yml/badge.svg)](https://github.com/gmile/strictly_speaking/actions/workflows/ci.yml)

Turns numbers into words. Similar to [humanize](https://github.com/radar/humanize).

## Installation

The package can be installed by adding `strictly_speaking` to your list of
dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:strictly_speaking, "~> 0.1"}
  ]
end
```

Documentation can be found at https://hexdocs.pm/strictly_speaking.

## Usage

```elixir
StrictlySpeaking.say(7, :ua)
# => "сім"

StrictlySpeaking.say(17, :en)
# => "seventeen"
```

## License

`strictly_speaking` is licensed under the MIT license, see the [LICENSE](LICENSE)
file.


CHANGELOG.md
# Changelog

## v0.1.1 (2021-03-20)

  * Add top-level module
  * Update README.md and add rudimentary documentation
  * Add support for pronunciation of 0 (zero)
  * Set up GitHub Actions CI

## v0.1.0 (2021-03-20)

  * Initial release


test/strictly_speaking_test.exs
defmodule StrictlySpeaking.Test do
  use ExUnit.Case

  doctest StrictlySpeaking

  describe "say/2" do
    test "raises on non-integer number" do
      assert_raise ArgumentError, "Number must be an integer. Found: 1.3", fn ->
        StrictlySpeaking.say(1.3, :ua)
      end
    end

    test "raises on unsupported language" do
      assert_raise ArgumentError, "Language 'th' is not implemented", fn ->
        StrictlySpeaking.say(14, :th)
      end
    end
  end
end


test/strictly_speaking/en_test.exs
defmodule StrictlySpeaking.EnTest do
  use ExUnit.Case

  doctest StrictlySpeaking.En
end


test/strictly_speaking/ua_test.exs
defmodule StrictlySpeaking.UaTest do
  use ExUnit.Case

  doctest StrictlySpeaking.Ua
end


lib/strictly_speaking.ex
defmodule StrictlySpeaking do
  @moduledoc """
  Implements a single function, `say/2`, which returns textual representation
  of a given integer in a given language.
  """

  @doc """
  Accepts an integer and a language symbol. Returns a string containing
  human-readable representation of given number.

  List of supported languages:

    * `:ua` - Ukrainian,
    * `:en` - English.

  Raises an ArgumentError in case `number` is not an integer,
  or `language` is not among supported.

  ## Examples

      iex> StrictlySpeaking.say(7, :ua)
      "сім"

      iex> StrictlySpeaking.say(17, :en)
      "seventeen"
  """
  def say(number, :en) when is_integer(number), do: StrictlySpeaking.En.say(number)
  def say(number, :ua) when is_integer(number), do: StrictlySpeaking.Ua.say(number)

  def say(number, language) when is_integer(number), do: raise(ArgumentError, "Language '#{language}' is not implemented")
  def say(number, _language), do: raise(ArgumentError, "Number must be an integer. Found: #{number}")
end


lib/strictly_speaking/ua.ex
defmodule StrictlySpeaking.Ua do
  @singles_m {"нуль", "один", "два", "три", "чотири", "п'ять", "шість", "сім", "вісім", "дев'ять"}
  @singles_f {"нуль", "одна", "двi", "три", "чотири", "п'ять", "шість", "сім", "вісім", "дев'ять"}

  @tens {"двадцять", "тридцять", "сорок", "п'ятдесят", "шістдесят", "сімдесят", "вісімдесят", "дев'яносто"}
  @teens {"десять", "одинадцять", "дванадцять", "тринадцять", "чотирнадцять", "п'ятнадцять", "шістнадцять", "сімнадцять", "вісімнадцять", "дев'ятндцять"}
  @hundreds {"сто", "двісті", "триста", "чотириста", "п'ятсот", "шістсот", "сімсот", "вісімсот", "дев'ятсот"}

  # TODO:
  #
  # квадрильйон
  # квінтильйон
  # секстильйон
  # септильйон
  # октильйон
  # нонильйон
  # децильйон
  @bigs {
    "тисяч",      "тисяча",   "тисячi",    "тисячi",    "тисячi",    "тисяч",      "тисяч",      "тисяч",      "тисяч",      "тисяч",
    "мільйонів",  "мільйон",  "мільйони",  "мільйони",  "мільйони",  "мільйонів",  "мільйонів",  "мільйонів",  "мільйонів",  "мільйонів",
    "мільярдів",  "мільярд",  "мільярди",  "мільярди",  "мільярди",  "мільярдів",  "мільярдів",  "мільярдів",  "мільярдів",  "мільярдів",
    "трильйонів", "трильйон", "трильйони", "трильйони", "трильйони", "трильйонів", "трильйонів", "трильйонів", "трильйонів", "трильйонів"
  }

  @doc """
  Accepts an integer. Returns a string containing human-readable representation of given number.

  ## Examples
      iex> StrictlySpeaking.Ua.say(0)
      "нуль"

      iex> StrictlySpeaking.Ua.say(7)
      "сім"

      iex> StrictlySpeaking.Ua.say(17)
      "сімнадцять"

      iex> StrictlySpeaking.Ua.say(23)
      "двадцять три"

      iex> StrictlySpeaking.Ua.say(117)
      "сто сімнадцять"

      iex> StrictlySpeaking.Ua.say(1_117)
      "одна тисяча сто сімнадцять"

      iex> StrictlySpeaking.Ua.say(10_117)
      "десять тисяч сто сімнадцять"

      iex> StrictlySpeaking.Ua.say(117_117)
      "сто сімнадцять тисяч сто сімнадцять"

      iex> StrictlySpeaking.Ua.say(123_456_789)
      "сто двадцять три мільйони чотириста п'ятдесят шість тисяч сімсот вісімдесят дев'ять"

  """
  def say(number, acc \\ << >>, order \\ 0)

  def say(0, _acc, _order), do: elem(@singles_m, 0)

  def say(number, acc, order) do
    {div1000, rem1000} = {div(number, 1000), rem(number, 1000)}
    {div100, rem100} = {div(rem1000, 100), rem(rem1000, 100)}
    {div10, rem10} = {div(rem100, 10), rem(rem100, 10)}

    sex = if order == 1, do: @singles_f, else: @singles_m

    result =
      case {order, div100, div10, rem10} do
        {_, 0, 0, 0} -> << >>
        {0, 0, 0, s} -> << elem(sex, s)::binary >>
        {0, 0, 1, s} -> << elem(@teens, s)::binary >>
        {0, 0, t, s} -> << elem(@tens, t - 2)::binary, ?\s, elem(sex, s)::binary >>

        {0, h, 0, 0} -> << elem(@hundreds, h - 1)::binary >>
        {0, h, 0, s} -> << elem(@hundreds, h - 1)::binary, ?\s, elem(sex, s)::binary >>
        {0, h, 1, s} -> << elem(@hundreds, h - 1)::binary, ?\s, elem(@teens, s)::binary >>
        {0, h, t, s} -> << elem(@hundreds, h - 1)::binary, ?\s, elem(@tens, t - 2)::binary, ?\s, elem(sex, s)::binary >>

        {o, 0, 0, s} -> << elem(sex, s)::binary, ?\s, elem(@bigs, 10 * (o - 1) + s)::binary >>
        {o, 0, 1, s} -> << elem(@teens, s)::binary, ?\s, elem(@bigs, 10 * (o - 1))::binary >>
        {o, 0, t, s} -> << elem(@tens, t - 1)::binary, ?\s, elem(sex, s)::binary, ?\s, elem(@bigs, 10 * (o - 1) + s)::binary >>

        {o, h, 0, 0} -> << elem(@hundreds, h - 1)::binary, ?\s, elem(@bigs, 10 * (o - 1))::binary >>
        {o, h, 0, s} -> << elem(@hundreds, h - 1)::binary, ?\s, elem(sex, s)::binary, ?\s, elem(@bigs, 10 * (o - 1) + s)::binary >>
        {o, h, 1, s} -> << elem(@hundreds, h - 1)::binary, ?\s, elem(@teens, s)::binary, ?\s, elem(@bigs, 10 * (o - 1))::binary >>
        {o, h, t, s} -> << elem(@hundreds, h - 1)::binary, ?\s, elem(@tens, t - 2)::binary, ?\s, elem(sex, s)::binary, ?\s, elem(@bigs, 10 * (o - 1) + s)::binary >>
      end

    case {result, div1000} do
      {<< >>,  div1000} -> say(div1000, acc, order + 1)
      {result,       0} -> << result::binary, acc::binary >>
      {result, div1000} -> say(div1000, << ?\s, result::binary, acc::binary >>, order + 1)
    end
  end
end


lib/strictly_speaking/en.ex
defmodule StrictlySpeaking.En do
  @singles {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}

  @tens {"twenty", "thirty", "fourty", "fifty", "sixty", "seventy", "eighty", "ninety"}
  @teens {"ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"}

  @bigs {"thousand", "million", "billion", "trillion"}

  @doc """
  Accepts an integer. Returns a string containing human-readable representation of given number.

  ## Examples
      iex> StrictlySpeaking.En.say(0)
      "zero"

      iex> StrictlySpeaking.En.say(7)
      "seven"

      iex> StrictlySpeaking.En.say(17)
      "seventeen"

      iex> StrictlySpeaking.En.say(23)
      "twenty three"

      iex> StrictlySpeaking.En.say(117)
      "one hundred and seventeen"

      iex> StrictlySpeaking.En.say(200)
      "two hundred"

      iex> StrictlySpeaking.En.say(217)
      "two hundred and seventeen"

      iex> StrictlySpeaking.En.say(227)
      "two hundred and twenty seven"

      iex> StrictlySpeaking.En.say(1_007)
      "one thousand seven"

      iex> StrictlySpeaking.En.say(1_017)
      "one thousand seventeen"

      iex> StrictlySpeaking.En.say(1_117)
      "one thousand one hundred and seventeen"

      iex> StrictlySpeaking.En.say(1_227)
      "one thousand two hundred and twenty seven"

      iex> StrictlySpeaking.En.say(17_117)
      "seventeen thousand one hundred and seventeen"

      iex> StrictlySpeaking.En.say(117_117)
      "one hundred and seventeen thousand one hundred and seventeen"

      iex> StrictlySpeaking.En.say(117_227)
      "one hundred and seventeen thousand two hundred and twenty seven"

  """
  def say(number, acc \\ << >>, order \\ 0)

  def say(0, _acc, _order), do: elem(@singles, 0)

  def say(number, acc, order) when number > 0 do
    {div1000, rem1000} = {div(number, 1000), rem(number, 1000)}
    {div100, rem100} = {div(rem1000, 100), rem(rem1000, 100)}
    {div10, rem10} = {div(rem100, 10), rem(rem100, 10)}

    result =
      case {order, div100, div10, rem10} do
        {_, 0, 0, 0} -> << >>
        {0, 0, 0, s} -> << elem(@singles, s)::binary >>
        {0, 0, 1, s} -> << elem(@teens, s)::binary >>
        {0, 0, t, s} -> << elem(@tens, t - 2)::binary, ?\s, elem(@singles, s)::binary >>

        {0, h, 0, 0} -> << elem(@singles, h)::binary, ?\s, "hundred" >>
        {0, h, 0, s} -> << elem(@singles, h)::binary, ?\s, "hundred", ?\s, "and", ?\s, elem(@singles, s)::binary >>
        {0, h, 1, s} -> << elem(@singles, h)::binary, ?\s, "hundred", ?\s, "and", ?\s, elem(@teens, s)::binary >>
        {0, h, t, s} -> << elem(@singles, h)::binary, ?\s, "hundred", ?\s, "and", ?\s, elem(@tens, t - 2)::binary, ?\s, elem(@singles, s)::binary >>

        {o, 0, 0, s} -> << elem(@singles, s)::binary, ?\s, elem(@bigs, o - 1)::binary >>
        {o, 0, 1, s} -> << elem(@teens, s)::binary, ?\s, elem(@bigs, o - 1)::binary >>
        {o, 0, t, s} -> << elem(@tens, t - 1)::binary, ?\s, elem(@singles, s)::binary, ?\s, elem(@bigs, o - 1)::binary >>

        {o, h, 0, 0} -> << elem(@singles, h)::binary, ?\s, "hundred", ?\s, elem(@bigs, o - 1)::binary >>
        {o, h, 0, s} -> << elem(@singles, h)::binary, ?\s, "hundred", ?\s, "and", ?\s, elem(@singles, s)::binary, ?\s, elem(@bigs, o - 1)::binary >>
        {o, h, 1, s} -> << elem(@singles, h)::binary, ?\s, "hundred", ?\s, "and", ?\s, elem(@teens, s)::binary, ?\s, elem(@bigs, o - 1)::binary >>
        {o, h, t, s} -> << elem(@singles, h)::binary, ?\s, "hundred", ?\s, "and", ?\s, elem(@tens, t - 2)::binary, ?\s, elem(@singles, s)::binary, ?\s, elem(@bigs, o - 1)::binary >>
      end

    case {result, div1000} do
      {<< >>,  div1000} -> say(div1000, acc, order + 1)
      {result,       0} -> << result::binary, acc::binary >>
      {result, div1000} -> say(div1000, << ?\s, result::binary, acc::binary >>, order + 1)
    end
  end
end

